cmd/tailscale/cli/up.go

```diff
@@ -603,30 +603,38 @@ func runUp(ctx context.Context, cmd string, args []string, upArgs upArgsT) (retE
 				msg := *n.ErrMessage
 				fatalf("backend error: %v\n", msg)
 			}
-			if s := n.State; s != nil {
-				switch *s {
-				case ipn.NeedsMachineAuth:
-					printed = true
-					if env.upArgs.json {
-						printUpDoneJSON(ipn.NeedsMachineAuth, "")
-					} else {
-						fmt.Fprintf(Stderr, "\nTo approve your machine, visit (as admin):\n\n\t%s\n\n", prefs.AdminPageURL(policyclient.Get()))
-					}
-				case ipn.Running:
-					// Done full authentication process
-					if env.upArgs.json {
-						printUpDoneJSON(ipn.Running, "")
-					} else if printed {
-						// Only need to print an update if we printed the "please click" message earlier.
-						fmt.Fprintf(Stderr, "Success.\n")
-					}
-					select {
-					case running <- true:
-					default:
-					}
-					cancelWatch()
+			if s := n.State; s != nil && *s == ipn.NeedsMachineAuth {
+				printed = true
+				if env.upArgs.json {
+					printUpDoneJSON(ipn.NeedsMachineAuth, "")
+				} else {
+					fmt.Fprintf(Stderr, "\nTo approve your machine, visit (as admin):\n\n\t%s\n\n", prefs.AdminPageURL(policyclient.Get()))
 				}
 			}
+
+			// If we've done the full authentication process, then we're done.
+			//
+			// If the user calls `tailscale up` without `--force-reauth`, this means
+			// waiting until ipn is in a running state.
+			//
+			// If the user `tailscale up --force-reauth`, then ipn may already be in
+			// a running state.  Wait until we got a "login finished" message from
+			// the backend.
```

Comment:
Does the "login finished" message only happen on force re-auth and not a regular `up`?

Reply:
I'm not sure, let me check.